examples/practical-coalton/15_pathnames.lisp

```diff
@@ -0,0 +1,177 @@
+;; To load this file, you need to load Chapter 15 from the Practical
+;; Common Lisp repository. The source code can be found at the link
+;; below. Then you can copy the Chapter 15 directory into
+;; ~/quicklisp/local-projects and run (quickload "Chapter-15").
+;; https://gigamonkeys.com/book/
+
+(in-package :cl-user)
+(ql:quickload "coalton")
+(ql:quickload "Chapter-15")
+(defpackage :practical-coalton.pathnames
+  (:use
+   #:coalton
+   #:coalton-prelude)
+  (:local-nicknames
+    (#:pth #:com.gigamonkeys.pathnames))
+  (:export
+   :list-directory
+   :file-exists-p
+   :directory-pathname-p
+   :file-pathname-p
+   :pathname-as-directory
+   :pathname-as-file
+   :walk-directory
+   :directory-p
+   :file-p))
+(in-package :practical-coalton.pathnames)
+
+(named-readtables:in-readtable coalton:coalton)
+
+;; Port of Chapter 15 from Practical Common Lisp, by Peter Seibel,
+;; to Coalton.
+;;
+;; In the previous chapter, we took the approach of rewriting the
+;; Common Lisp functions in Coalton. In this chapter, we instead
+;; write a thin wrapper around the Common Lisp functions.
+;;
+;; First, we start by defining a few Coalton types that provide
+;; a type-safe interface to the Common Lisp pathname functions.
+;; For example, in several of the Common Lisp functions a wildcard
+;; pathname is an invalid argument. Instead of checking for wildcard
+;; paths and erroring if found, we can use the type system to ensure
+;; that we never pass a wildcard path to a function that doesn't
+;; accept it.
+;;
+;; Second, we define a corresponding Coalton function for each
+;; function exported by the Common Lisp pathname package.
+
+;; TODO: I'm open to any suggestions to make this code less verbose.
```

Comment:
I'm open to any suggestions to make this code less verbose.

If you count the comments at the top of the file and the package definition (which the original doesn't have), this is almost as long as the Common Lisp implementation this is wrapping!